lib/api/v3/utilities/custom_field_injector.rb

```diff
@@ -0,0 +1,223 @@
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+module API
+  module V3
+    module Utilities
+      class CustomFieldInjector
+        TYPE_MAP = {
+          'string' => 'String',
+          'text' => 'Formattable',
+          'int' => 'Integer',
+          'float' => 'Float',
+          'date' => 'Date',
+          'bool' => 'Boolean',
+          'user' => 'User',
+          'version' => 'Version',
+          'list' => 'StringObject'
+        }
+
+        LINK_FORMATS = ['list', 'user', 'version']
+
+        PATH_METHOD_MAP = {
+          'user' => :user,
+          'version' => :version,
+          'list' => :string_object
+        }
+
+        NAMESPACE_MAP = {
+          'user' => 'users',
+          'version' => 'versions',
+          'list' => 'string_objects'
+        }
+
+        class << self
+          def linked_field?(custom_field)
+            LINK_FORMATS.include?(custom_field.field_format)
+          end
+
+          def property_field?(custom_field)
+            !linked_field?(custom_field)
+          end
+        end
+
+        def initialize(representer_class)
+          @class = representer_class
+        end
+
+        # N.B. accepting a wp_schema here is not too great, but seems like the best way
+        # to obtain available versions and users for WP custom fields
+        def inject_schema(custom_field, wp_schema: nil)
+          case custom_field.field_format
+          when 'version'
+            inject_version_schema(custom_field, wp_schema)
+          when 'user'
+            inject_user_schema(custom_field, wp_schema)
+          when 'list'
+            inject_list_schema(custom_field)
+          else
+            inject_basic_schema(custom_field)
+          end
+        end
+
+        def inject_value(custom_field)
+          case custom_field.field_format
+          when *LINK_FORMATS
+            inject_link_value(custom_field)
+          else
+            inject_property_value(custom_field)
+          end
+        end
+
+        def inject_patchable_link_value(custom_field)
```

Comment:
Assignment Branch Condition size for inject_patchable_link_value is too high. [29.56/15]<br>Perceived complexity for inject_patchable_link_value is too high. [9/7]